src/miral/idle_listener.cpp

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "miral/idle_listener.h"
+#include <mir/server.h>
+#include <mir/shell/idle_handler.h>
+
+class miral::IdleListener::Impl : public mir::shell::IdleHandlerObserver
+{
+public:
+    explicit Impl(IdleListenerParams const& params)
+        : on_dim{params.on_dim},
+          on_off{params.on_off},
+          on_wake{params.on_wake}
```

Comment:
I think I'd move the tests for empty `function<>`s here:
```suggestion
        : on_dim{params ? params.on_dim : []{}},
          on_off{params ? params.params.on_off : []{}},
          on_wake{params ? params.params.on_wake : []{}}
```

(Why is there no `std::function<>::else(...)`?)

Reply:
`IdleListenerParams` now has default empty callbacks